scripts/test/test_upload_all.py

```diff
@@ -0,0 +1,180 @@
+from pathlib import Path
+from typing import Dict
+from unittest import mock
+from unittest.mock import Mock, call
+
+import pytest
+
+from upload_all import (
+    glob_output_files,
+    Args,
+    parse_args,
+    upload_images_to_openstack,
+    upload_single_image,
+    main,
+    get_upload_name,
+)
+
+
+def test_glob_output_files(tmp_path):
+    """
+    Test that the glob_output_files function returns the expected files
+    and does not include directories
+    :param tmp_path: A temp path setup by pytest
+    """
+    expected_output_path = tmp_path / "k8s-image-builder" / "images" / "capi" / "output"
+    expected_output_path.mkdir(parents=True)
+
+    expected = ["ubuntu-kube-1.23.4", "ubuntu-kube-1.40.0"]
+    for out_name in expected:
+        # Make parent dir as this is a named dir with a single file
+        (expected_output_path / out_name).mkdir(parents=True)
+        (expected_output_path / out_name / out_name).touch()
+        # And a fake file in each
+        (expected_output_path / out_name / "should_not_appear").touch()
+
+    with mock.patch("git.Repo") as mocked_repo:
+        mocked_repo.return_value.working_tree_dir = str(tmp_path)
+        returned = list(glob_output_files())  # Avoid exhausting the generator
+
+    expected = sorted(str(expected_output_path / file / file) for file in expected)
+    assert sorted(str(file) for file in returned) == expected
+    assert len(returned) == len(expected)
+
+
+def test_arg_parse_all_true():
+    """
+    Tests that the parse_args function returns the expected Args object
+    if the user sets all optional arguments
+    """
+    user_input = ["--dry-run", "--public", "default"]
+    expected = Args(dry_run=True, public=True, os_cloud="default")
+    assert parse_args(user_input) == expected
+
+
+def test_arg_parse_no_optional_args():
+    """
+    Tests that the parse_args function returns the expected Args object
+    if the user does not set any optional arguments
+    """
+    user_input = ["bar"]
+    expected = Args(dry_run=False, public=False, os_cloud="bar")
+    assert parse_args(user_input) == expected
+
+
+def test_arg_parse_no_args():
+    """
+    Tests that the parse_args function exits with a SystemExit
+    if the user does not set the OS cloud argument
+    """
+    user_input = []
+    with pytest.raises(SystemExit):
+        parse_args(user_input)
+
+
+def test_upload_images_to_openstack_dry_run():
+    """
+    Tests that the upload_images_to_openstack function does not call
+    anything uploading to OpenStack if the dry_run flag is set
+    """
+    expected_files = [Path("file1"), Path("file2")]
+    args = Args(dry_run=True, public=True, os_cloud="default")
+
+    with mock.patch("upload_all.upload_single_image") as mocked_upload_single_image:
+        upload_images_to_openstack(expected_files, args)
+    assert mocked_upload_single_image.call_count == 0
+
+
+def _expected_args_helper(file: Path, visibility: str) -> Dict:
```

Comment:
shouldn't this be a fixture?

Reply:
Yeah I keep forgetting we have pytest :man_facepalming:

Reply:
Ah I remember why

In `test_upload_images_to_openstack_real_run` I want to simulate two images being created, the problem is the struct holds the filename for each.

This means I need to generate two fixture objects with different paths or simply use list comprehension and a factory pattern like we did here:
```
    expected = [
        call(args.os_cloud, _expected_args_helper(file, "public"))
        for file in expected_files
    ]
```

Do you know of an easy way to use fixtures with a pattern a little bit like:
```
def test_upload_images_to_openstack_real_run(fixture_1, fixture_2, [Path(file_1), Path(file_2)]):
```

Since most of the solutions I can see multiple the boilerplate?